File: Java/1201.java
__________________________________________________________________________________________________
sample 0 ms submission
class Solution {
    public int nthUglyNumber(int n, int a, int b, int c) {
        int min = 1, max = 2_000_000_001;
        while(min < max){
            int mid = min + (max - min) / 2;
            if (isFloor(n, a, b, c, mid)){
                min = mid + 1;
            } else {
                max = mid;
            }
        }
        return max;
    }

    public boolean isFloor(int n, int a, int b, int c, int mid){
        int count = mid / a + mid / b + mid / c;
        count = (int) (count - mid / maxCommonDivisor(a, b) - mid / maxCommonDivisor(a, c) - mid / maxCommonDivisor(b, c) + mid / maxCommonDivisor(maxCommonDivisor(a, b), c));
        return count < n;
    }

    private long maxCommonDivisor(long m, long n) {
        long tmp = m * n;
        if (m < n) {
            long temp = m;
            m = n;
            n = temp;
        }
        while (m % n != 0) {
            long temp = m % n;
            m = n;
            n = temp;
        }
        return tmp / n;
    }
}
__________________________________________________________________________________________________
sample 1 ms submission
class Solution {
    public int nthUglyNumber(int n, int a, int b, int c) {
        long low = 1, high = Integer.MAX_VALUE, mid;  
  
        while (low < high) {  
            mid = low + (high - low) / 2;  
  
            // If the current term is less than  
            // n then we need to increase low  
            // to mid + 1  
            if (divTermCount(a, b, c, mid) < n)  
                low = mid + 1;  
      
            // If current term is greater than equal to  
            // n then high = mid  
            else
                high = mid;  
        }  
  
        return (int)low;
    }
    static long divTermCount(long a, long b, long c, long num)  
    {  
  
        // Calculate number of terms divisible by a and  
        // by b and by c then, remove the terms which is are  
        // divisible by both a and b, both b and c, both  
        // c and a and then add which are divisible by a and  
        // b and c  
        return ((num / a) + (num / b) + (num / c)  
                - (num / lcm(a, b))  
                - (num / lcm(b, c))  
                - (num / lcm(a, c))  
                + (num / lcm(a, lcm(b, c))));  
    }
    static long gcd(long a, long b)  
    {  
        if (a == 0)  
            return b;  
          
        return gcd(b % a, a);  
    }  
  
    // Function to return the lcm of a and b  
    static long lcm(long a, long b)  
    {  
        return (a * b) / gcd(a, b);  
    }  
}
__________________________________________________________________________________________________

File: Python3/1201.py
__________________________________________________________________________________________________
sample 36 ms submission
import functools, operator
class Solution:
    def nthUglyNumber(self, n: int, a: int, b: int, c: int) -> int:
        # this is a binary search
        # 2 3 3
        # 2 3 4 6 8
        lo, hi = 1, 2 * 1000000000
        # factors = list({a, b, c})
        # factors.sort()
        # print(factors)
        a, b, c = sorted([a, b, c])
        
        def count(num):
            return num // a + num // b + num // c - num // (a * b // math.gcd(a, b)) - num // (a * c // math.gcd(a, c)) - num // (b * c // math.gcd(b, c)) + num // (a * b * c // (max( math.gcd(a, b) * math.gcd(b, c), math.gcd(a, c) * math.gcd(b, c), math.gcd(a, b) * math.gcd(a, c)) )  )
        
        while lo < hi:
            mid = (lo + hi) >> 1
            cnt = count(mid)
            if cnt < n:
                lo = mid + 1
            else:
                hi = mid
        return lo
        
__________________________________________________________________________________________________
sample 44 ms submission
class Solution:
    def nthUglyNumber(self, n: int, a: int, b: int, c: int) -> int:
        from math import gcd
        
        def count(num):
            return num // a + num // b + num // c - num // ab - num // ac - num // bc + num // abc
        
        ab, ac, bc = a * b // gcd(a, b), a * c // gcd(a, c), b * c // gcd(b, c)
        abc = ab * c // gcd(ab, c)
        lo, hi = 1, 2 * 10 ** 9
        while lo < hi:
            mi = (lo + hi) // 2
            if count(mi) >= n:
                hi = mi
            else:
                lo = mi + 1
        return lo
__________________________________________________________________________________________________
